### PR TITLE
Use records start & end date instead of session (if records available)

### PR DESCRIPTION
**Pull request by thomaschampagne**
This PR is triggered by a bug encountered on the following fit export: [RGTCycling_Cap-Formentor_2020-01-09_17-49-14.zip](https://github.com/sports-alliance/sports-lib/files/4241682/RGTCycling_Cap-Formentor_2020-01-09_17-49-14.zip)

The start & end date were equivalent (=> `duration = 0`) and in future (year 2126):

image

I also experienced the issue on others files exported by others users. Strava handles the fit file properly:

image

The PR uses the first and last timestamp of records (if available) to fix this case.

---

**Important:** The added test fails because the power stream length (no null or infinity filtering) should have a size of **1746** instead of **1745**. All others streams (not squashed) have a length of **1746**. Any tips so i can fix it :)?

### CODE REVIEW

**Comment by jimmykane** [comment]
Hey, thanks for this. 

I ll take a look to this in the week (how to solve the power sample size)

**Comment by jimmykane** [comment]
Looking at the utilities for the sample size this is done: 

```
  public static getDataLength(startDate: Date, endDate: Date): number {
    return Math.ceil((+endDate - +startDate) / 1000);
  }
```

So gets the date diff in miliseconds and then finds the seconds using `ceil` so end date is inclusive to the second (used a sample slug).

**Comment by jimmykane** [comment]
Which I dont think it's well done. I am adding tests now. Let me work a bit on this.

**Comment by jimmykane** [comment]
I added some changed and made `1746 ` the official size as it should be.

File: src/events/adapters/importers/fit/importer.fit.mapper.ts [hunk 1]
```diff
@@ -106,7 +106,7 @@ export const FITSampleMapper: { dataType: string, getSampleValue(sample: any): n
   {
     dataType: DataPower.type,
     getSampleValue: (sample: any) => {
-      return sample.power || sample.Power;
+      return isNumber(sample.power) ? sample.power : sample.Power;
```
Comment (jimmykane) [inline src/events/adapters/importers/fit/importer.fit.mapper.ts]: @thomaschampagne here was the issue. I think the last sample was 0 and typically JS returned false

Comment (thomaschampagne) [reply src/events/adapters/importers/fit/importer.fit.mapper.ts]: Nice found! Perfect :)

**Comment by thomaschampagne** [comment]
There's another issue with the duration.  Currently duration is zero. It should be `1745` seconds.

New test assertion: `expect(event.getDuration().getValue()).toEqual(1745);`

We could get it like this:

```typescript
private static getDurationStats(activity: Activity): DataInterface[] {

    const stats = [];
    
    if (activity.endDate && activity.startDate) {
      const elapsedTime = activity.endDate.getTime() - activity.startDate.getTime();
      stats.push(new DataDuration(elapsedTime));
    }

    return stats;
}
```
The pause time is also impacted in the legacy code:

```typescript
  private static getStatsFromObject(object: any): DataInterface[] {
    ....
    ....
    const pause = (object.total_elapsed_time > totalTimerTime ?
      object.total_elapsed_time - totalTimerTime :
      totalTimerTime - object.total_elapsed_time) || 0;
    stats.push(new DataPause(pause));
```

I identified various duration & pause time problems in PR https://github.com/sports-alliance/sports-lib/pull/43. 

You can merge this PR without fixing duration & pause time. I will update PR https://github.com/sports-alliance/sports-lib/pull/43 with develop, add new assertion and provide my fixes for duration & pause time for all formats.

**Comment by jimmykane** [comment]
If you need 1745 check the reverted commit. It's a debatable situation.
Floor vs ceil.

Let me commit in a bit when baby sleeps to debate on what we prefer.

**Comment by thomaschampagne** [comment]
Thanks for the merge :) When i spoke of `1745` in my last comment. I was talking about time: duration of the activty which should be `1745` seconds instead of `0`. Below the same test with the duretion assertion:

image

**Comment by jimmykane** [comment]
Could we add this test to the other PR so I can work on this ?

**Comment by thomaschampagne** [comment]
@jimmykane Sure let me perform the rebase of develop on https://github.com/sports-alliance/sports-lib/pull/43 + add assertion.  You will be notified in https://github.com/sports-alliance/sports-lib/pull/43

From my analysis  of strava & garmin platforms the concept of  "is in movement or not" is different for every sports. I mean the speeds thresholds to detect movement is not the same for all sports.

- For cycling, move/rest is detected around an average of  1.8 m/s
- For running, move/rest is detected around an average of  1m/s

I think this depends of the grade adjusted speed (it should...).

I will explain all of this again in https://github.com/sports-alliance/sports-lib/pull/43

**Pull request merged**